File: WordPatter.java
// Time Complexity O(n) - n number of patterns
// Space Complexity O(1) - as there can be max 26 characters else the condition will break
class Solution {
    public boolean wordPattern(String pattern, String s) {
        HashMap map = new HashMap<>();
        
         String[] str = s.split(" ");
        
        if(str.length != pattern.length()) return false;
        
        for(Integer i =0; i<str.length; i++) {
            Character character = pattern.charAt(i);
            String word = str[i];
               
            if(!map.containsKey(character)) {
                map.put(character, i);
            }
            
            if(!map.containsKey(word)) {
                map.put(word, i);
            }
            
            if(map.get(character) != map.get(word)) return false;
            
        }
        
        return true;
    }
}

File: WordPatter.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        d = {}
        l = s.split(" ")

        if len(l) != len(pattern):
            return False 

        for idx, ch in enumerate(pattern):
            if ch not in d.keys() and l[idx] not in d.values():
                d[ch] = l[idx]
            elif ch not in d.keys():
                return False
            elif d[ch] != l[idx]:
                return False
        return True